db/db_impl/db_impl.h

```diff
@@ -2063,6 +2064,15 @@ class DBImpl : public DB {
   InstrumentedCondVar atomic_flush_install_cv_;
 
   bool wal_in_db_path_;
+
+  // Next notification id for completion listeners.
+  uint64_t next_notification_id_;
+  // Completion listeners queue.
+  std::deque<uint64_t> notification_queue_;
```

Comment:
We need a extra lock for keeping notification in sequence, since hold mutex_ is not expectable during notifying listeners.
The original idea of maintaining a queue is we can have some parallelism for parparing the object for notificaiton.
e.g. :
1. nlock.lock()
2. Add notification id into queue.
3. nlock.unlock()
4. prepare notification object <-- parallel tasks
5. wait con_var for current id is the first one of the queue.
6. send notifications
But it looks like it's a little bit overkill, considering the extra code complexity. I have simplified the logic.

Reply:
Yeah, I totally get the need for the mutex, was just wondering about the queue (since it was FIFO apparently). And I do agree about the complexity (it's better to keep it simple).

Reply:
@burtonli Actually, question: isn't causality ensured simply by issuing the notification before scheduling further work?

Reply:
Oh, or is it that *another* background task might also finish around the same time and schedule a compaction?

Reply:
Correct, in the original logic, there is no lock to protect flush or compaction completion notification, even we can make sure current thread doesn't trigger any new compaction before notification, but there may be other background tasks jump ahead. It's easy to get a repro by setting aggressive LSM compaction, e.g.; 10KB memtable size, L0 compaction trigger = 4.

Reply:
Makes sense, thanks for clarifying!